lib/ecto/type.ex

```diff
@@ -301,92 +296,124 @@ defmodule Ecto.Type do
       iex> dump({:array, :binary}, ["1", "2", "3"])
       {:ok, ["1", "2", "3"]}
 
-  A `dumper` function may be given for handling recursive types.
+  A `dumper` function may be given for handling recursive types, see: `dump/3`.
   """
-  @spec dump(t, term, (t, term -> {:ok, term} | :error)) :: {:ok, term} | :error
-  def dump(type, value, dumper \\ &dump/2)
-
-  def dump(_type, nil, _dumper) do
+  @spec dump(t, term) :: {:ok, term} | :error
+  def dump(_type, nil) do
     {:ok, nil}
   end
 
-  def dump(:binary_id, value, _dumper) when is_binary(value) do
-    {:ok, value}
+  def dump(type, value) do
+    dump_fun(type).(value)
   end
 
-  def dump(:any, value, _dumper) do
-    {:ok, value}
+  @doc """
+  Dumps a value to the given type using a custom function.
```

Comment:
I think this paragraph is not accurate, otherwise I would expect the catch-rest function to be like:

```elixir
def dump(_type, value, dumper) do
  dumper.(value)
end
```

(which makes no sense because you could just use the `dumper` function directly)